File: website/extract-data.py
#!/usr/bin/env python3
import os
import json
import re
from pathlib import Path

class DataExtractor:
    def __init__(self):
        self.base_path = Path(__file__).parent.parent
        self.exercises_path = self.base_path / 'Exercises'
        self.website_path = self.base_path / 'website'
    
    def extract_all_data(self):
        exercises = []
        exercise_dirs = sorted([d for d in os.listdir(self.exercises_path) 
                               if d.startswith('Exercise')])
        
        for exercise_dir in exercise_dirs:
            exercise_data = self.extract_exercise_data(exercise_dir)
            exercises.append(exercise_data)
        
        # Save extracted data
        data_path = self.website_path / 'data'
        data_path.mkdir(exist_ok=True)
        
        with open(data_path / 'exercises.json', 'w', encoding='utf-8') as f:
            json.dump(exercises, f, ensure_ascii=False, indent=2)
        
        print(f"Extracted data for {len(exercises)} exercises")
        return exercises
    
    def extract_exercise_data(self, exercise_dir):
        exercise_path = self.exercises_path / exercise_dir
        readme_path = exercise_path / 'README.md'
        solutions_path = exercise_path / 'Solutions'
        
        exercise_data = {
            'id': exercise_dir.lower(),
            'directory': exercise_dir,
            'title': '',
            'difficulty': 1,
            'description': '',
            'readme': '',
            'solutions': []
        }
        
        # Read README.md
        if readme_path.exists():
            with open(readme_path, 'r', encoding='utf-8') as f:
                exercise_data['readme'] = f.read()
            
            # Extract title and difficulty from README
            title_match = re.search(r'^# (.+)', exercise_data['readme'], re.MULTILINE)
            if title_match:
                exercise_data['title'] = title_match.group(1)
                
                # Extract difficulty
                difficulty_match = re.search(r'難易度:\s*(\d+)', exercise_data['title'])
                if difficulty_match:
                    exercise_data['difficulty'] = int(difficulty_match.group(1))
            
            # Extract description from overview section
            overview_match = re.search(r'## 概要\n(.*?)\n\n', exercise_data['readme'], re.DOTALL)
            if overview_match:
                exercise_data['description'] = overview_match.group(1).strip()
        
        # Read solution files
        if solutions_path.exists():
            solution_files = sorted([f for f in os.listdir(solutions_path) 
                                   if f.endswith('.cs')])
            
            for solution_file in solution_files:
                solution_path = solutions_path / solution_file
                with open(solution_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                exercise_data['solutions'].append({
                    'filename': solution_file,
                    'content': content
                })
        
        return exercise_data

if __name__ == '__main__':
    extractor = DataExtractor()
    extractor.extract_all_data()

File: website/extract-data.js
// Data loader script to extract content from the repository files
const fs = require('fs');
const path = require('path');

class DataExtractor {
    constructor() {
        this.basePath = path.join(__dirname, '..');
        this.exercisesPath = path.join(this.basePath, 'Exercises');
        this.websitePath = path.join(this.basePath, 'website');
    }

    async extractAllData() {
        const exercises = [];
        const exerciseDirs = fs.readdirSync(this.exercisesPath)
            .filter(dir => dir.startsWith('Exercise'))
            .sort();

        for (const exerciseDir of exerciseDirs) {
            const exerciseData = await this.extractExerciseData(exerciseDir);
            exercises.push(exerciseData);
        }

        // Save extracted data
        const dataPath = path.join(this.websitePath, 'data');
        if (!fs.existsSync(dataPath)) {
            fs.mkdirSync(dataPath, { recursive: true });
        }

        fs.writeFileSync(
            path.join(dataPath, 'exercises.json'),
            JSON.stringify(exercises, null, 2),
            'utf8'
        );

        console.log(`Extracted data for ${exercises.length} exercises`);
        return exercises;
    }

    async extractExerciseData(exerciseDir) {
        const exercisePath = path.join(this.exercisesPath, exerciseDir);
        const readmePath = path.join(exercisePath, 'README.md');
        const solutionsPath = path.join(exercisePath, 'Solutions');

        const exerciseData = {
            id: exerciseDir.toLowerCase(),
            directory: exerciseDir,
            title: '',
            difficulty: 1,
            description: '',
            readme: '',
            solutions: []
        };

        // Read README.md
        if (fs.existsSync(readmePath)) {
            exerciseData.readme = fs.readFileSync(readmePath, 'utf8');
            
            // Extract title and difficulty from README
            const titleMatch = exerciseData.readme.match(/^# (.+)/);
            if (titleMatch) {
                exerciseData.title = titleMatch[1];
                
                // Extract difficulty
                const difficultyMatch = exerciseData.title.match(/難易度:\s*(\d+)/);
                if (difficultyMatch) {
                    exerciseData.difficulty = parseInt(difficultyMatch[1]);
                }
            }

            // Extract description from overview section
            const overviewMatch = exerciseData.readme.match(/## 概要\n(.*?)\n\n/s);
            if (overviewMatch) {
                exerciseData.description = overviewMatch[1].trim();
            }
        }

        // Read solution files
        if (fs.existsSync(solutionsPath)) {
            const solutionFiles = fs.readdirSync(solutionsPath)
                .filter(file => file.endsWith('.cs'))
                .sort();

            for (const solutionFile of solutionFiles) {
                const solutionPath = path.join(solutionsPath, solutionFile);
                const content = fs.readFileSync(solutionPath, 'utf8');
                
                exerciseData.solutions.push({
                    filename: solutionFile,
                    content: content
                });
            }
        }

        return exerciseData;
    }
}

// Run if this script is executed directly
if (require.main === module) {
    const extractor = new DataExtractor();
    extractor.extractAllData().catch(console.error);
}

module.exports = DataExtractor;